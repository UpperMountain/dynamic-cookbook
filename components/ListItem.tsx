import React from "react";
import {
  StyleSheet,
  View,
  ViewProps,
  Image,
  Text,
  TouchableOpacity
} from "react-native";
import { Asset } from "expo";

const styles = StyleSheet.create({
  item: {
    display: "flex",
    flexDirection: "row"
  },
  imageContainer: {
    width: 50,
    height: 50,
    borderRadius: 25,
    backgroundColor: "rgba(0,0,0,0.05)",
    overflow: "hidden"
  },
  image: {
    width: 50,
    height: 50
  },
  body: {
    flex: 1,
    marginLeft: 15
  },
  bodyText: {
    height: 50,
    display: "flex",
    flexDirection: "column",
    justifyContent: "center"
  }
});

export default function ListItem({
  name,
  body,
  children,
  image,
  onPress,
  style
}: {
  name: string;
  body?: string;
  image?: Asset;
  children?: React.ReactNode;
  onPress?: () => void;
  style?: ViewProps["style"];
}) {
  const contents = (
    <>
      <View style={styles.imageContainer}>
        {image && <Image style={styles.image} source={image} />}
      </View>
      <View style={styles.body}>
        <View style={styles.bodyText}>
          <Text numberOfLines={1}>{name}</Text>
          {body && (
            <Text numberOfLines={1} style={{ color: "rgba(0,0,0,0.55)" }}>
              {body}
            </Text>
          )}
        </View>
        {children}
      </View>
    </>
  );

  // add touch handler only if enabled
  if (onPress) {
    return (
      <TouchableOpacity onPress={onPress} style={[styles.item, style]}>
        {contents}
      </TouchableOpacity>
    );
  } else {
    return <View style={[styles.item, style]}>{contents}</View>;
  }
}
