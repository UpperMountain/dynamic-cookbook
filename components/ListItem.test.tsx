import React from "react";
import ListItem from "./ListItem";
import renderer from "react-test-renderer";

it("should render without exploding", () => {
  renderer.create(<ListItem name="test name" body="test body" />).unmount();
});

it("should pass snapshot tests", () => {
  const cmp = renderer.create(
    <>
      <ListItem name="test name" body="test body" />
      <ListItem name="nested outer">
        <ListItem name="nested inner" />
      </ListItem>
      <ListItem
        name="with image"
        body="ooooohh pretty"
        image={require("../assets/images/CookieLogo.png")}
      />
      <ListItem name="with press handler" onPress={jest.fn()} />
      <ListItem name="with custom style" style={{ backgroundColor: "red" }} />
    </>
  );
  expect(cmp.toJSON()).toMatchSnapshot();
  cmp.unmount();
});
