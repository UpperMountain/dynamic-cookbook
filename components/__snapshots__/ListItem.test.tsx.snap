// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass snapshot tests 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.05)",
          "borderRadius": 25,
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 15,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
        >
          test name
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(0,0,0,0.55)",
            }
          }
        >
          test body
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.05)",
          "borderRadius": 25,
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 15,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
        >
          nested outer
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.05)",
              "borderRadius": 25,
              "height": 50,
              "overflow": "hidden",
              "width": 50,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "marginLeft": 15,
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
                "height": 50,
                "justifyContent": "center",
              }
            }
          >
            <Text
              numberOfLines={1}
            >
              nested inner
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.05)",
          "borderRadius": 25,
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 15,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
        >
          with image
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(0,0,0,0.55)",
            }
          }
        >
          ooooohh pretty
        </Text>
      </View>
    </View>
  </View>,
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.05)",
          "borderRadius": 25,
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 15,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
        >
          with press handler
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Object {
          "backgroundColor": "red",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.05)",
          "borderRadius": 25,
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 15,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          numberOfLines={1}
        >
          with custom style
        </Text>
      </View>
    </View>
  </View>,
]
`;
